File: packages/core-ts/src/reliance/status-list.ts
// SPDX-License-Identifier: Apache-2.0
// Credential status for the experimental RM v1 binding, per W3C Bitstring Status
// List v1.0: `encodedList` is the multibase base64url (prefix "u", no padding)
// encoding of the GZIP-compressed bitstring; bit 0 is the most significant bit of
// the first byte. Decompression is bounded. The status list is itself a protected
// artifact and counts only when signed by the target credential's own issuer.
// Node-only (zlib); not exported from the browser-reachable barrel.
import { gunzipSync, gzipSync } from 'node:zlib';
import type { SemanticState } from './types.js';

/** Minimum bitstring length in bits (Bitstring Status List v1.0). */
export const MIN_STATUS_BITS = 131_072;
/** Upper bound for a decompressed list (16 Mbit), so a small list cannot inflate unboundedly. */
export const MAX_STATUS_BYTES = 2 * 1024 * 1024;

export class StatusListError extends Error {
  constructor(public readonly code: 'MALFORMED' | 'TOO_LARGE' | 'TOO_SHORT', message: string) {
    super(message);
    this.name = 'StatusListError';
  }
}

/** Encode a bitstring as a Bitstring Status List `encodedList` value. */
export function encodeStatusList(bits: Uint8Array): string {
  if (bits.byteLength * 8 < MIN_STATUS_BITS) {
    throw new StatusListError('TOO_SHORT', `A status list needs at least ${MIN_STATUS_BITS} bits.`);
  }
  // Fixed header fields (mtime 0) keep the output deterministic.
  return `u${Buffer.from(gzipSync(bits, { level: 9 })).toString('base64url')}`;
}

/** Decode an `encodedList` value with bounded decompression. */
export function decodeStatusList(encoded: unknown, maxBytes = MAX_STATUS_BYTES): Uint8Array {
  if (typeof encoded !== 'string' || !/^u[A-Za-z0-9_-]+$/.test(encoded)) {
    throw new StatusListError('MALFORMED', 'encodedList must be multibase base64url (prefix "u").');
  }
  let bits: Buffer;
  try {
    bits = gunzipSync(Buffer.from(encoded.slice(1), 'base64url'), { maxOutputLength: maxBytes });
  } catch (error) {
    const code = (error as { code?: string }).code;
    if (code === 'ERR_BUFFER_TOO_LARGE' || /maxOutputLength|buffer/i.test(String(error))) {
      throw new StatusListError('TOO_LARGE', `Decompressed status list exceeds ${maxBytes} bytes.`);
    }
    throw new StatusListError('MALFORMED', 'encodedList is not valid GZIP data.');
  }
  if (bits.byteLength * 8 < MIN_STATUS_BITS) {
    throw new StatusListError('TOO_SHORT', `Status list has fewer than ${MIN_STATUS_BITS} bits.`);
  }
  return new Uint8Array(bits);
}

/** Read bit `index` (most significant bit first). */
export function statusBit(bits: Uint8Array, index: number): boolean {
  const byte = bits[Math.floor(index / 8)];
  if (byte === undefined) throw new RangeError(`Status index ${index} is outside the list.`);
  return ((byte >> (7 - (index % 8))) & 1) === 1;
}

/** Verifier-owned status rules taken from the reliance profile. */
export interface StatusPolicy {
  readonly required: boolean;
  readonly purposes: readonly string[];
  readonly maxAgeSeconds: number;
}

export interface StatusOutcome {
  readonly state: SemanticState;
  readonly reason: string;
  readonly listUri?: string;
  readonly sources: readonly string[];
}

function isObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

/**
 * Evaluate the target's `credentialStatus` against an already verified status-list
 * credential. `list` is undefined when the list could not be resolved; `listProtected`
 * reports whether the list's own protection and validity were established.
 * Unavailable, unauthorized or stale status is not_established; an applicable set bit
 * is contradicted.
 */
export function evaluateStatus(
  credential: Record<string, unknown>,
  list: Record<string, unknown> | undefined,
  listProtected: SemanticState,
  policy: StatusPolicy,
  evaluationTime: string,
): StatusOutcome {
  const entry = credential.credentialStatus;
  if (entry === undefined) {
    return policy.required
      ? { state: 'not_established', reason: 'Status is required by the profile but the credential names none.', sources: [] }
      : { state: 'established', reason: 'The profile does not require status for this credential.', sources: [] };
  }
  if (!isObject(entry) || entry.type !== 'BitstringStatusListEntry' ||
      typeof entry.statusListCredential !== 'string' || typeof entry.statusListIndex !== 'string') {
    return { state: 'not_established', reason: 'Unsupported credentialStatus form.', sources: ['/credentialStatus'] };
  }
  const listUri = entry.statusListCredential;
  const sources = ['/credentialStatus', listUri];
  if (!policy.purposes.includes(String(entry.statusPurpose))) {
    return { state: 'not_established', reason: `Status purpose ${String(entry.statusPurpose)} is not accepted by the profile.`, listUri, sources };
  }
  if (list === undefined) {
    return { state: 'not_established', reason: `Status list ${listUri} is unavailable.`, listUri, sources };
  }
  if (listProtected !== 'established') {
    return { state: 'not_established', reason: `Status list ${listUri} is not protected and valid.`, listUri, sources };
  }
  if (list.id !== listUri) {
    return { state: 'not_established', reason: `Resolved status list identifies itself as ${String(list.id)}.`, listUri, sources };
  }
  if (list.issuer !== credential.issuer) {
    return { state: 'not_established', reason: `Status list is signed by ${String(list.issuer)}, who may not state status for credentials of ${String(credential.issuer)}.`, listUri, sources };
  }
  const subject = list.credentialSubject;
  if (!isObject(subject) || subject.type !== 'BitstringStatusList' || subject.statusPurpose !== entry.statusPurpose) {
    return { state: 'not_established', reason: 'Status list purpose does not match the entry.', listUri, sources };
  }
  const at = Date.parse(evaluationTime), issued = Date.parse(String(list.validFrom));
  if (!Number.isFinite(at) || !Number.isFinite(issued) || (at - issued) / 1000 > policy.maxAgeSeconds) {
    return { state: 'not_established', reason: `Status list is older than the profile's ${policy.maxAgeSeconds} s freshness limit.`, listUri, sources };
  }
  if (!/^(0|[1-9][0-9]*)$/.test(entry.statusListIndex)) {
    return { state: 'not_established', reason: 'statusListIndex is not a non-negative integer.', listUri, sources };
  }
  let revoked: boolean;
  try {
    revoked = statusBit(decodeStatusList(subject.encodedList), Number(entry.statusListIndex));
  } catch (error) {
    return { state: 'not_established', reason: `Status list cannot be read: ${(error as Error).message}`, listUri, sources };
  }
  return revoked
    ? { state: 'contradicted', reason: `Revoked: bit ${entry.statusListIndex} of ${listUri} is set.`, listUri, sources }
    : { state: 'established', reason: `Not revoked: bit ${entry.statusListIndex} of ${listUri} is clear.`, listUri, sources };
}

File: packages/core-py/qi_vc_core/reliance/status_list.py
# SPDX-License-Identifier: Apache-2.0
"""Credential status for the experimental RM v1 binding (mirrors status-list.ts).

W3C Bitstring Status List v1.0: encodedList is multibase base64url (prefix "u", no
padding) of the GZIP-compressed bitstring, bit 0 = most significant bit of byte 0.
Decompression is bounded. A list counts only when signed by the credential's issuer.
"""

from __future__ import annotations

import base64
import gzip
import re
import zlib
from dataclasses import dataclass
from datetime import datetime
from typing import Any, Literal

from .types import SemanticState

MIN_STATUS_BITS = 131_072
MAX_STATUS_BYTES = 2 * 1024 * 1024
_ENCODED = re.compile(r"^u[A-Za-z0-9_-]+$")


class StatusListError(ValueError):
    def __init__(
        self, code: Literal["MALFORMED", "TOO_LARGE", "TOO_SHORT"], message: str
    ) -> None:
        super().__init__(message)
        self.code = code


def encode_status_list(bits: bytes) -> str:
    """Encode a bitstring as an encodedList value (GZIP, mtime 0, multibase "u")."""
    if len(bits) * 8 < MIN_STATUS_BITS:
        raise StatusListError(
            "TOO_SHORT", f"A status list needs at least {MIN_STATUS_BITS} bits."
        )
    compressed = gzip.compress(bits, compresslevel=9, mtime=0)
    return "u" + base64.urlsafe_b64encode(compressed).decode("ascii").rstrip("=")


def decode_status_list(encoded: Any, max_bytes: int = MAX_STATUS_BYTES) -> bytes:
    """Decode an encodedList value with bounded decompression."""
    if not isinstance(encoded, str) or not _ENCODED.fullmatch(encoded):
        raise StatusListError(
            "MALFORMED", 'encodedList must be multibase base64url (prefix "u").'
        )
    body = encoded[1:]
    try:
        compressed = base64.urlsafe_b64decode(body + "=" * (-len(body) % 4))
        inflater = zlib.decompressobj(wbits=31)  # GZIP container only
        bits = inflater.decompress(compressed, max_bytes)
        if inflater.unconsumed_tail:
            raise StatusListError(
                "TOO_LARGE", f"Decompressed status list exceeds {max_bytes} bytes."
            )
        if not inflater.eof:
            raise StatusListError("MALFORMED", "encodedList is not valid GZIP data.")
    except (ValueError, zlib.error) as error:
        if isinstance(error, StatusListError):
            raise
        raise StatusListError(
            "MALFORMED", "encodedList is not valid GZIP data."
        ) from error
    if len(bits) * 8 < MIN_STATUS_BITS:
        raise StatusListError(
            "TOO_SHORT", f"Status list has fewer than {MIN_STATUS_BITS} bits."
        )
    return bits


def status_bit(bits: bytes, index: int) -> bool:
    if index // 8 >= len(bits):
        raise IndexError(f"Status index {index} is outside the list.")
    return bool((bits[index // 8] >> (7 - index % 8)) & 1)


@dataclass(frozen=True)
class StatusPolicy:
    required: bool
    purposes: tuple[str, ...]
    max_age_seconds: int


@dataclass(frozen=True)
class StatusOutcome:
    state: SemanticState
    reason: str
    sources: tuple[str, ...]
    list_uri: str | None = None


def _time(value: Any) -> datetime | None:
    if not isinstance(value, str):
        return None
    try:
        return datetime.fromisoformat(
            value[:-1] + "+00:00" if value.endswith("Z") else value
        )
    except ValueError:
        return None


def select_status_entry(
    credential: dict[str, Any], purposes: tuple[str, ...] | list[str]
) -> tuple[dict[str, Any] | None, str | None]:
    """The single status entry for one of ``purposes`` (mirrors selectStatusEntry).

    Entries for other purposes are left to whoever evaluates them; several entries
    for the requested purposes are ambiguous. Returns (entry, reason); both None
    when the credential names no status.
    """
    value = credential.get("credentialStatus")
    if value is None:
        return None, None
    entries = value if isinstance(value, list) else [value]
    if not entries or not all(isinstance(entry, dict) for entry in entries):
        return None, "Unsupported credentialStatus form."
    matching = [e for e in entries if str(e.get("statusPurpose")) in purposes]
    if len(matching) > 1:
        return None, f"Several status entries for {'/'.join(purposes)}."
    if not matching:
        found = ", ".join(str(e.get("statusPurpose")) for e in entries)
        return None, (
            f"No status entry has an accepted purpose ({', '.join(purposes)}); "
            f"found {found}."
        )
    return matching[0], None


def evaluate_status(
    credential: dict[str, Any],
    status_list: dict[str, Any] | None,
    list_protected: SemanticState,
    policy: StatusPolicy,
    evaluation_time: str,
) -> StatusOutcome:
    """Unavailable, unauthorized or stale status is not_established.

    An applicable set bit is contradicted.
    """
    entry, problem = select_status_entry(credential, policy.purposes)
    if entry is None and problem is None:
        if policy.required:
            return StatusOutcome(
                "not_established",
                "Status is required by the profile but the credential names none.",
                (),
            )
        return StatusOutcome(
            "established",
            "The profile does not require status for this credential.",
            (),
        )
    if entry is None:
        return StatusOutcome("not_established", str(problem), ("/credentialStatus",))
    if (
        entry.get("type") != "BitstringStatusListEntry"
        or not isinstance(entry.get("statusListCredential"), str)
        or not isinstance(entry.get("statusListIndex"), str)
    ):
        return StatusOutcome(
            "not_established",
            "Unsupported credentialStatus form.",
            ("/credentialStatus",),
        )
    uri: str = entry["statusListCredential"]
    sources = ("/credentialStatus", uri)

    def no(reason: str) -> StatusOutcome:
        return StatusOutcome("not_established", reason, sources, uri)

    if status_list is None:
        return no(f"Status list {uri} is unavailable.")
    if list_protected != "established":
        return no(f"Status list {uri} is not protected and valid.")
    if status_list.get("id") != uri:
        return no(f"Resolved status list identifies itself as {status_list.get('id')}.")
    if status_list.get("issuer") != credential.get("issuer"):
        return no(
            f"Status list is signed by {status_list.get('issuer')}, who may not state "
            f"status for credentials of {credential.get('issuer')}."
        )
    subject = status_list.get("credentialSubject")
    if (
        not isinstance(subject, dict)
        or subject.get("type") != "BitstringStatusList"
        or subject.get("statusPurpose") != entry.get("statusPurpose")
    ):
        return no("Status list purpose does not match the entry.")
    at, issued = _time(evaluation_time), _time(status_list.get("validFrom"))
    if (
        at is None
        or issued is None
        or (at - issued).total_seconds() > policy.max_age_seconds
    ):
        return no(
            f"Status list is older than the profile's {policy.max_age_seconds} s "
            "freshness limit."
        )
    index = entry["statusListIndex"]
    if not re.fullmatch(r"0|[1-9][0-9]*", index):
        return no("statusListIndex is not a non-negative integer.")
    try:
        is_set = status_bit(decode_status_list(subject.get("encodedList")), int(index))
    except (StatusListError, IndexError) as error:
        return no(f"Status list cannot be read: {error}")
    yes, not_ = (
        ("Suspended", "Not suspended")
        if entry.get("statusPurpose") == "suspension"
        else ("Revoked", "Not revoked")
    )
    if is_set:
        return StatusOutcome(
            "contradicted", f"{yes}: bit {index} of {uri} is set.", sources, uri
        )
    return StatusOutcome(
        "established", f"{not_}: bit {index} of {uri} is clear.", sources, uri
    )
